api/src/main/java/org/apache/iceberg/expressions/StrictMetricsEvaluator.java

```diff
@@ -392,7 +392,12 @@ public Boolean or(Boolean leftResult, Boolean rightResult) {
       Types.NestedField field = struct.field(id);
       Preconditions.checkNotNull(field, "Cannot filter by nested column: %s", schema.findField(id));
 
-      if (containsNullsOnly(id) || containsNaNsOnly(id)) {
+      if (canContainNulls(id)) {
+        // NOT null IN (1, 2) -> NOT null -> null -> false in filters
+        return ROWS_MIGHT_NOT_MATCH;
+      }
+
+      if (containsNaNsOnly(id)) {
```

Comment:
I have a question about NaN semantics. Do we consider `NaN == NaN`? Looks like not. I think most databases consider NaN values equivalent even though it is against the standard. I guess Spark follows Postgres.